src/Elasticsearch.Extensions.Logging/ElasticsearchDataProcessor.cs

```diff
@@ -0,0 +1,263 @@
+
+			return _ecs;
+		}
+
+		internal Agent GetAgent()
+		{
+			if (_agent is null)
+			{
+				var assembly = typeof(ElasticsearchLogger).Assembly;
+				var type = assembly.GetName().Name;
+				var versionAttribute = assembly.GetCustomAttributes(false)
+					.OfType<AssemblyInformationalVersionAttribute>()
+					.FirstOrDefault();
+				var version = versionAttribute?.InformationalVersion;
+				_agent = new Agent { Type = type, Version = version };
+			}
+
+			return _agent;
+		}
+
+		internal Host GetHost()
+		{
+			if (_host is null)
+			{
+				// Architecture osArchitecture = RuntimeInformation.OSArchitecture;
+				// if (osDescription.Contains('#'))
+				// {
+				//     int indexOfHash = osDescription.IndexOf('#');
+				//     osDescription = osDescription.Substring(0, Math.Max(0, indexOfHash - 1));
+				// }
+
+				var operatingSystem = new Os
+				{
+					Full = RuntimeInformation.OSDescription,
+					Platform = Environment.OSVersion.Platform.ToString(),
+					Version = Environment.OSVersion.Version.ToString()
+				};
+				_host = new Host
+				{
+					Hostname = Environment.MachineName, Architecture = RuntimeInformation.OSArchitecture.ToString(), Os = operatingSystem
+				};
+			}
+
+			return _host;
+		}
+
+		internal Process GetProcess()
+		{
+			if (_processName == null)
+			{
+				using var process = System.Diagnostics.Process.GetCurrentProcess();
+				_processId = process.Id;
+				_processName = process.ProcessName;
+			}
+
+			var currentThread = Thread.CurrentThread;
+
+			return new Process
+			{
+				Name = _processName,
+				Pid = _processId,
+				Thread = new ProcessThread { Name = currentThread.Name, Id = currentThread.ManagedThreadId }
+			};
+		}
+
+		internal Service GetService()
+		{
+			if (_service is null)
+			{
+				var entryAssembly = Assembly.GetEntryAssembly();
+				var entryAssemblyName = entryAssembly.GetName();
+				var type = entryAssemblyName.Name;
+				var versionAttribute = entryAssembly.GetCustomAttributes(false)
+					.OfType<AssemblyInformationalVersionAttribute>()
+					.FirstOrDefault();
+				var version = versionAttribute?.InformationalVersion ?? entryAssemblyName.Version.ToString();
+				_service = new Service { Type = type, Version = version };
+			}
+
+			return _service;
+		}
+
+		internal int GetSeverity(LogLevel logLevel)
+		{
+			switch (logLevel)
+			{
+				case LogLevel.Critical:
+					return 2;
+				case LogLevel.Error:
+					return 3;
+				case LogLevel.Warning:
+					return 4;
+				case LogLevel.Information:
+					return 6;
+				case LogLevel.Debug:
+				case LogLevel.Trace:
+					return 7;
+				default:
+					return 7;
+			}
+		}
+
+		private void PostEvent(LogEvent logEvent)
+		{
+			var indexTime = logEvent.Timestamp ?? ElasticsearchLoggerProvider.LocalDateTimeProvider();
+			if (_options.IndexOffset.HasValue) indexTime = indexTime.ToOffset(_options.IndexOffset.Value);
+
+			var index = string.Format(_options.Index, indexTime);
+
+			var id = Guid.NewGuid().ToString();
```

Comment:
If the id for a document is auto-generated, it'd be better to not send an id and rely on Elasticsearch to generate one. This is faster in Elasticsearch because it does not need to check if a document already exists with a given id.